### ISSUE TITLE
Add cursor based pagination (GraphQL connections)

### ISSUE THREAD

**Issue by asishallab** [Zendro-dev/graphql-server-model-codegen#44]
Read [this document](https://graphql.org/learn/pagination/) to understand why we need cursor based pagination.

Implement it in both the GraphQL-server as well the single page application.

**Comment by asishallab** [comment]
### A very brief summary of what to do

Currently pagination accepts two classical argument: `limit` and `offset`
Here, the above `offset` defines from where on records will be fetched and returned, e.g. `offset: 100` means `ignore the first hundred`.

Switch to `cursor` where `cursor: 100` means `return after ID:100`.

**Comment by asishallab** [comment]
### Deviation from the standard GraphQL connection scheme

Our prime reason why we need cursor based pagination is explained in slides eight to ten [in this presentation](https://docs.google.com/presentation/d/1fIMq25GsqLFKqS4-vdqxlwWHWn9er_iA26CVNgmvrLw/edit#slide=id.g4a15d75193_0_12).

As documented [here](https://graphql.org/learn/pagination/) GraphQL suggests the usage of a `connection`, which is in fact a complex query and return-object to be defined in the respective GraphQL schema. For our purposes we only want to adopt the usage of a `cursor` instead of an `offset`. We do not want to return the complex `connection` object as for now. For some inspiration and doubts about the GraphQL connection approach see [this GitHub conversation](https://github.com/graphql/graphql-relay-js/issues/27)

This means that we need to change the `paginate` functions generated to not expect an `offset` argument, but a `cursor` argument. Note that the new `first` argument is just the renamed old `limit` argument. 

_Question_: How to transform a `cursor` into an `offset`?

The cursor is the complete last object base64 encoded. It thus can be decoded and used for ordering purposes and enables identification of the next batch. Consider the following pseudo code:

```
# Dog model and its existing records:
[
 { name: 'Old yeller', age: 12, id: 1},
 { name: 'Old yeller', age: 14, id: 2},
 { name: 'Chiquito', age: 8, id: 3}
]

# Current cursor is e.g.
cursor = base64( { name: 'Old yeller', age: 12, id: 1} )

# Then assume a GraphQL query is send with the above `cursor` and `after: 2`
# The paginate function needs to analyze the arguments provided to the `order` function, if any.

# Analyze the arguments of order(...) an for each ordered attribute _i_ do
# If desc -> comparator[i] will be '<'
# else comparator[i] will be '>'

# Based on these data the following SQL query can be constructed.
SELECT * FROM dogs WHERE dog.name <comparator[i]> <cursorObj.name> AND dog.id <comparator[i]> <cursorObj.id> ORDER BY dog.name, dog.id

# Note, that the above comparators have to be added for each attribute appearing in the arguments of `order(...)` and that finally the `id` has to be added as the last order attribute. 
```

**Comment by asishallab** [comment]
### How to use just the `ID` attribute as `cursor` and skip `base64` encoding?

If we just send the value of the `ID` attribute of the last record in the current batch as the `cursor`, we could obtain the associated record by a simple fetch from the database. Consider the above pseudo code and the following modification of it:

```
cursor = currentDog.id

cursorObj = models.dog.readById( cursor )

# continue as above
```

**Comment by asishallab** [comment]
So far:
```
books {
  search( args ) {
    return [ JavaSCript Plain Old Objects ]
  }

  toBase64() {
    return base64( this.withoutAssocs )
  }
}

person {
  hasManyBooks
  
  getBooks() {
    return books.search( {person_id: this.id} )
  }

  toBase64() {
    return base64( this.withoutAssocs )
  }
}

// If we change to full connection model
// The above would break

person {
  // ...
  getBooks() {
    return books.search{ {person_id: this.id} ).edges.map(x => { return x.node})
  }
}

// If we want to avoid this, we could add simply a 
// NEW function (resolver and query) that return a GraphQL-connection 
// (https://graphql.org/learn/pagination/#pagination-and-edges)

// Define additional and new cursorPaginationArg to
// (https://graphql.org/learn/pagination/#pagination-and-edges)
{
  after: // cursor base64 encoded (If NULL first page, else the base64-encoded 'last' record of the current page)
  first: // limit - how many records to show per page (page-size)
}

books {
  search( args ) {
    // stays the same
  }

  // new connection-function (resolver)
  graphQlConnection(searchArgs, connectionPaginationArg, orderArgs) {
    // use the already implemented functions search, paginate, order
    // and create a connection object that is returned
    return {
      totalCount
      edges: [
      { // an edge
        node {
          // the book
        }
        cursor: // the book base64 encoded WITHOUT associations
      }, 
      // ...
      ],
      pageInfo: {
        endCursor
        hasNextPage
      }
   }

   // In DETAIL:
   // First parse the arguments and obtain the requested records
   // Consider the ordering as described in the Issue
   records = searchAndOrderAndPaginateLikeDescribedInTheAboveCodeCommentAndIssueAndPutSomeMoreTextToConfuseEveryoneOk?()
   // Now construct the "edges"
   edges = transformRecordArrayToEdgesArr
   
   // Function can be a helper that looks like this:
   function transformRecordArrayToEdgesArr(recordsArr) {
     return recordsArr.map( x => { return { node: x, cursor: x.toBase64() } )
   }
}
```

**Comment by vsuaste** [comment]
Commits [#](https://github.com/ScienceDb/graphql-server-model-codegen/commit/e66152d77578f6d00b35352efe22bfb4a2f7decf) [#ea6f28](https://github.com/ScienceDb/graphql-server-model-codegen/commit/ea6f285e47d477fd386705c6a73eff3851c68f31) [#81bf33](https://github.com/ScienceDb/graphql-server-model-codegen/commit/81bf33d4d343eb806aa948ed808d071dad4dd8a7) [#cfc96b](https://github.com/ScienceDb/graphql-server-model-codegen/commit/cfc96ba09319bb79969ca8c858c8c1bf040889cc) [#5db726](https://github.com/ScienceDb/graphql-server-model-codegen/commit/5db726a50955d99619f68c2f61cab1bf3cf252e9) [#c5dae7](https://github.com/ScienceDb/graphql-server-model-codegen/commit/c5dae77e584a2472a068453ed6bd28c6853f8b98) [#324181](https://github.com/ScienceDb/graphql-server-model-codegen/commit/324181f73426295845f0fda7ff9e5a4f9758a96f) [#0e2af3](https://github.com/ScienceDb/graphql-server-model-codegen/commit/0e2af36c9abe6a9a16146396351e36891580f4e0) [#4fd257](https://github.com/ScienceDb/graphql-server-model-codegen/commit/4fd257a23922d1172a0474230c7797c3bbbde893) [#213602](https://github.com/ScienceDb/graphql-server-model-codegen/commit/213602bf784ea7769231456987f4ab79bf92017d)  [#d4e560](https://github.com/ScienceDb/graphql-server-model-codegen/commit/d4e5607e9fa76d06f789248842d26f3f1e65415f)

**Comment by framirez07** [comment]
Reopened to add support for **backward** cursor based pagination.

**Comment by framirez07** [comment]
New commits: 

on **graphql-server-model-codegen**:
[i44 - [in progress] - backward cursor based pagination support added](https://github.com/ScienceDb/graphql-server-model-codegen/commit/eda704463833301a248af62c6e5a80d1324c7e6b)

on **graphql-server**:
[i44 - [in progress] - backward cursor based pagination support added](https://github.com/ScienceDb/graphql-server/commit/abe421fbf2ff1d1294d9705baa1e56b000e15d5f)

**Comment by asishallab** [comment]
A few final issues remain to be tackled:

1. backward pagination
2. refresh, i.e. including the cursor 

Basis for the following definitions is [Facebook Relay's proposed standard](https://facebook.github.io/relay/graphql/connections.htm).

_Note_ the usage of two cursor arguments `after` and `before`. They could be merged into a single `cursor` argument, which'd fulfill the [KISS](https://en.wikipedia.org/wiki/KISS_principle), but alas we aim to keep with Facebook's proposed standard.

### Backward pagination

#### Validation of function-arguments

In the model implementation validate the provided arguments and raise an error, if invalid, i.e. non-sense arguments are provided.
```
let argsValid = (pagination === undefined) || (pagination.first && !pagination.before && !pagination.last) || (pagination.last && !pagination.after && !pagination.first)
if (!argsValid) {
  throw new Error('Illlegal cursor based pagination arguments. ...')
}
```

#### Decide whether forward or backward pagination is requested by the user

```
let isForwardPagination = (pagination === undefined || pagination.first)
// or even simpler 
let isBackwardPagination = (pagination.last != undefined)
```

### Enable refresh of current page

To enable refreshing the current displayed page, we need to enable including the cursor. Add an _optional_ argument `includeCursor` to the schema, resolvers, and model API. If and only if `includeCursor` is `true` the current cursor (`after` or `before`) is included and thus the "offset arguments" `last` or `first` will be diminished by one to ensure the resulting number of records is equal to the respective offset argument. This is a tricky bit, because theoretically the cursor record might have been deleted during the time elapsed before the refresh. In that case, the pagination should not diminish the offset argument and work exactly as if `includeCursor` was `false`.

**Comment by asishallab** [comment]
### Pagination when NOT sorting by id

Consider the following example, where a data model has the attributes `id` and `name` and the user requests sorting by `name. We show, how to ensure correct result sets.
```
{id: 4, name: 'A'}
{id: 3, name: 'A'}
{id: 2, name: 'B'}
{id: 5, name: 'C'}

// cursor id=3
WHERE name > 'A' OR name = 'A' and id > 3 ORDER BY name, id
// include cursor - note "id >= 3"
WHERE name > 'A' OR name = 'A' and id >= 3 ORDER BY name, id
```

**Comment by coeit** [comment]
The following function (sketch) should solve the problem (Thanks @asishallab ):
```
function(order, cursor) {
  // assume id is present among order attributes
  // assume more than one order argument
  // assume id is always last in the order arguments
  let lastFields = []
  let sqlQuery = ''
  for (i = 0; i < order.length; i++) {
    let curr = order[i]
    let currField = curr[0]
    let currOp = curr[1] === 'ASC' ? '>' : '<'
    let currSql = `${currField} ${currOp} '${c[currField]}'`
    let lastFieldSql = false
    if (lastFields.length > 0) {
      lastFieldSql = lastFields.map( f => `${f} = '${cursor[f]}'`).join(' AND ')
    }
    lastFields.push(currField)
    if (sqlQuery.length > 0) {
      sqlQuery += ' OR '
    }
    sqlQuery += lastFieldSql ? `${lastFieldSql} AND ${currSql}` : currSql
    //return as sequelize 'where' argument (options['where'])
  }
  return sqlQuery
}
``` 
### remaining concerns

1. The function generates SQL and not Sequelize where objects. That would need to be adopted accordingly, or maybe we should investigate if providing our where directly as SQL would work?  
2. The integer value for id is put in between ' and appears as string (see [...] AND id > '1')
3. The function makes three assumptions that need to be fulfilled(see sketch).
4. We need to verify that the idAttribute is present among the order arguments. If not, we should push it into the last position of the order argument Array.
5. The function is not able to deal with backwards pagination.
6. The function needs to be able to deal with the (to be implemented) `includeCursor` parameter to include the record associated with the given cursor in the results
